### DISCUSSION TITLE
Git에서 들이면 좋은 습관 :)

### DISCUSSION THREAD

**Discussion by designe** [euphony-io/euphony#100] [Knowledge Share]
git에서 들이면 좋은 습관을 한번 공유드려보려고 합니다.

간단한데요.

바로 master에서는 작업을 하지 않는 것입니다.
자기 자신의 master는 Pull Request를 날리고자 하는 org와 fetch upstream을 통해 맞춰두는게 좋구요.

새로운 피쳐를 추가할때는 브랜치를 활용하는 것이 가장 깔끔하게 contribution할 수 있는 방법인데요.
예를 들어 utf16패치를 하고자 할때에는 아래와 같이 master로부터 branch를 만듭니다.

```
git checkout -b support-utf16
```

그리고 여기서 utf16 관련된 commit을 이 브랜치에서 작업후 commit & push 합니다.
그리고 euphony-io의 master로 pull request를 날리게 되면 깔끔하게 이전 이력이 남지 않게 됩니다.

그리고 또 다른 피쳐를 만들고자 할때는 master로부터 브랜치를 또 갈라서
```
git checkout master
git checkout -b new-feature
```
이런식으로 피쳐 단위(pull request로 이력 남기고 싶은)로 브랜치를 나누는게
간단하지만 좋은 습관이기 때문에 공유하게 되었습니다 :)

PS.
최근에 Pull Request를 받아보면 간혹 이미 merge가 된 커밋들인데
PR 이력에 그대로 남는 경우를 보셨죠?!
자세히 보면 euphony-io org의 커밋들의 commit id가 다를거에요.
이유는 정말 다른 커밋이기 때문입니다! 
?!
euphony-io에서 커밋의 최종 merge되는 커밋은 squash를 통해
1PR 당 1commit으로 이뤄집니다.
하나의 커밋만 발생시켰다 하더라도 마지막 merge때 수정사항(메세지라도)이 발생하면
다른 커밋으로 판단됩니다. 내용은 똑같더라도 ㅠ
참고해주세요!! :)

**Comment by kuro11pow2** [comment]
squash가 PR을 발생시킨 브랜치에는 영향을 주지 않는 거였군요. 다시 생각해보니 그냥 두 브랜치 사이의 squash merge일 뿐이니까 영향을 주지 않는 게 당연하다는 생각이 드네요! 자세한 설명 감사드립니다 👍

**Reply by designe** [reply]
:)

**Reply by designe** [reply]
+1

**Comment by jopopcorn** [comment]
> 최근에 Pull Request를 받아보면 간혹 이미 merge가 된 커밋들인데
> PR 이력에 그대로 남는 경우를 보셨죠?!
> 자세히 보면 euphony-io org의 커밋들의 commit id가 다를거에요.
> 이유는 정말 다른 커밋이기 때문입니다!

오오 PS에 적어주신 이 내용은 처음 알았네요 좋은 팁 공유해주셔서 감사합니다!

**Comment by sujin9** [comment]
저도 PS 부분은 처음 알았네요!! 꿀팁 감사합니다 😊

**Comment by judemin** [comment]
이건 앱과는 조금 거리가 먼 질문인데
여러 브랜치에서 작업할 경우, 웹 프레임워크를 사용하면 
.gitignore에 node_module이나 .env등 환경 변수 파일들을 넣는 경우가 많았습니다.
이때 브랜치를 여러개 사용하면 ignore된 파일이 사라지거나 파일이 꼬이는 경우가 있었는데
이렇게 파일 등이 꼬이는 경우 멘토님들께서는 어떻게 해결하시는지 궁금합니다.

**Reply by yeonns** [reply]
가끔 gitignore가 적용되지 않는 경우를 겪어보긴 했는데, 왠만하면 캐시 삭제로 해결됐던것같아요!
https://stackoverflow.com/questions/11451535/gitignore-is-ignored-by-git
```sh
git rm -r --cached .
git add .
git commit -m "fixed untracked files"
```

**Comment by aiclaudev** [comment]
깃은 친해지고싶어도 참 힘든 것 같습니다.. 말씀해주신 팁 토대로 구글링하면서 더 공부해야할거같습니다ㅠㅠㅠ 좋은 지식 공유 감사합니다!!👍👍

**Comment by psh2849** [comment]
주로 master에서 작업했었는데 이런 문제가 있었네요 감사합니다 👍 👍

**Comment by dlwlstks96** [comment]
소소하지만 결과물이 훨씬 깔끔해지는 아주 좋은 습관이 될 것 같습니다!! 좋은 정보 감사합니다:)

**Comment by seohy02** [comment]
"이미 merge가 된 커밋들인데 PR 이력에 그대로 남는 경우" 이런 경우를 많이 봐서 궁금했었는데...!! 감사합니다!!

**Reply by Jsueeee** [reply]
저도 이 부분이 인상적이네요 ㅎㅎㅎ

**Comment by ycs1m1yk** [comment]
좋은 브랜칭 습관과 pr 이력 지식공유 감사합니다!
